mllib/src/main/scala/org/apache/spark/ml/evaluation/RankingMetrics.scala

```diff
@@ -0,0 +1,202 @@
+        }
+        cnt.toDouble / k
+      } else {
+        logWarning("Empty ground truth set, check input data")
+        0.0
+      }
+    }, DoubleType)
+
+    val predictionColumn: Column = predictionAndObservations.col(predictionCol)
+    val labelColumn: Column = predictionAndObservations.col(labelCol)
+
+    predictionAndObservations
+      .withColumn("predictionAtK", precisionAtK(predictionColumn, labelColumn))
+      .agg(mean("predictionAtK")).first().getDouble(0)
+  }
+
+  /**
+   * Returns the mean average precision (MAP) of all the queries.
+   * If a query has an empty ground truth set, the average precision will be zero and a log
+   * warning is generated.
+   */
+  lazy val meanAveragePrecision: Double = {
+
+    def map = udf((predicted: Seq[Any], actual: Seq[Any]) => {
+      val actualSet = actual.toSet
+      if (actualSet.nonEmpty) {
+        var i = 0
+        var cnt = 0
+        var precSum = 0.0
+        val n = predicted.length
+        while (i < n) {
+          if (actualSet.contains(predicted(i))) {
+            cnt += 1
+            precSum += cnt.toDouble / (i + 1)
+          }
+          i += 1
+        }
+        precSum / actualSet.size
+      } else {
+        logWarning("Empty ground truth set, check input data")
+        0.0
+      }
+    }, DoubleType)
+
+    val predictionColumn: Column = predictionAndObservations.col(predictionCol)
+    val labelColumn: Column = predictionAndObservations.col(labelCol)
+
+    predictionAndObservations
+      .withColumn("MAP", map(predictionColumn, labelColumn))
+      .agg(mean("MAP")).first().getDouble(0)
+  }
+
+  /**
+   * Compute the average NDCG value of all the queries, truncated at ranking position k.
+   * The discounted cumulative gain at position k is computed as:
+   *    sum,,i=1,,^k^ (2^{relevance of ''i''th item}^ - 1) / log(i + 1),
+   * and the NDCG is obtained by dividing the DCG value on the ground truth set. In the current
+   * implementation, the relevance value is binary.
+
+   * If a query has an empty ground truth set, zero will be used as ndcg together with
+   * a log warning.
+   *
+   * See the following paper for detail:
+   *
+   * IR evaluation methods for retrieving highly relevant documents. K. Jarvelin and J. Kekalainen
+   *
+   * @param k the position to compute the truncated ndcg, must be positive
+   * @return the average ndcg at the first k ranking positions
+   */
+  @Since("2.2.0")
+  def ndcgAt(k: Int): Double = {
+    require(k > 0, "ranking position k should be positive")
+
+    def ndcgAtK = udf((predicted: Seq[Any], actual: Seq[Any]) => {
+      val actualSet = actual.toSet
+
+      if (actualSet.nonEmpty) {
+        val labSetSize = actualSet.size
+        val n = math.min(math.max(predicted.length, labSetSize), k)
+        var maxDcg = 0.0
+        var dcg = 0.0
+        var i = 0
+        while (i < n) {
+          val gain = 1.0 / math.log(i + 2)
+          if (i < predicted.length && actualSet.contains(predicted(i))) {
```

Comment:
This doesn't seem right, there is no overlap between the calculation of dcg and max_dcg. The question asked here should be if the label at predicted(i) is "good". When treating the labels as binary relevant/not relevant I suppose that might use a threshold, but better would be to move away from a binary dcg and use the full equation from the docblock. I understand though that you are not looking to make major updates to the code from mllib, so it would probably be reasonable for someone to fix this in a followup.